### PR TITLE
BUG: fix segfault on side-effect in __bool__ function in array.nonzero()

### PR DESCRIPTION
**Pull request by garyg1**
Numpy should throw an exception here, instead of segfaulting. See #13631

### CODE REVIEW

File: numpy/core/src/multiarray/item_selection.c [hunk 1]
```diff
@@ -2289,8 +2289,15 @@ PyArray_Nonzero(PyArrayObject *self)
         }
         else {
             npy_intp j;
+            npy_intp added_count = 0;
             for (j = 0; j < count; ++j) {
                 if (nonzero(data, self)) {
+                    if (++added_count > nonzero_count) {
+                        PyErr_SetString(PyExc_RuntimeError,
```
Comment (seberg) [inline numpy/core/src/multiarray/item_selection.c]: I do not believe you can set the errorstate without the GIL? You should probably just break and check if something odd happened after the fact. If we error at too many items read, I think we may as well error at too few items read.

File: numpy/core/tests/test_item_selection.py [hunk 2]
```diff
@@ -79,9 +79,29 @@ def test_empty_partition(self):
...
+
+        b = np.array([0, 1, 2, 3, 4, 5])
+        assert_array_equal(a, b)
+
+    def test_nonzero_safety(self):
+        # from Github isue #13631
+        class FalseThenTrue:
+            _val = False
+            def __bool__(self):
+                try: return self._val
```
Comment (seberg) [inline numpy/core/tests/test_item_selection.py]: Please put the new line here, I do not think we use this style anywhere else

File: numpy/core/tests/test_item_selection.py [hunk 3]
```diff
@@ -79,9 +79,29 @@ def test_empty_partition(self):
...
+                finally: self._val = True
+
+        try:
+            np.array([True, FalseThenTrue()]).nonzero()
+            assert_(False)
```
Comment (seberg) [inline numpy/core/tests/test_item_selection.py]: Probably might as well use `assert` nowadays, not sure I prefer `pass` instead of `assert True`.

**Review by seberg — COMMENTED** [review]
I guess we may sometimes miss if the output has is empty (no True values), that seems fine, it just uses a single pass then, which really is also OK behaviour.

File: numpy/core/tests/test_item_selection.py (hunk 3 above)
Comment (eric-wieser) [reply numpy/core/tests/test_item_selection.py]: This should be using
```
with pytest.assert_raises(RuntimeError):
    np.array([True, FalseThenTrue()]).nonzero()
```

File: numpy/core/tests/test_item_selection.py [hunk 4]
```diff
@@ -79,9 +79,63 @@ def test_empty_partition(self):
...
-            a = a.argpartition(np.array([], dtype=np.int16))
+        # In reference to github issue #6530
+        a = np.array([0, 2, 4, 6, 8, 10])
+        a = a.argpartition(np.array([], dtype=np.int16))
+
+        b = np.array([0, 1, 2, 3, 4, 5])
+        assert_array_equal(a, b)
+
+    def test_nonzero_safety(self):
+        # from Github isue #13631
```
Comment (eric-wieser) [inline numpy/core/tests/test_item_selection.py]: `gh-13631` is how we typically spell this

File: numpy/core/tests/test_item_selection.py [hunk 5]
```diff
@@ -79,9 +79,63 @@ def test_empty_partition(self):
...
-            a = np.array([0, 2, 4, 6, 8, 10])
-            a = a.argpartition(np.array([], dtype=np.int16))
+        # In reference to github issue #6530
+        a = np.array([0, 2, 4, 6, 8, 10])
+        a = a.argpartition(np.array([], dtype=np.int16))
+
+        b = np.array([0, 1, 2, 3, 4, 5])
+        assert_array_equal(a, b)
+
+    def test_nonzero_safety(self):
```
Comment (eric-wieser) [inline numpy/core/tests/test_item_selection.py]: Though I can see why you put it here, this test should go in `core/tests/test_numeric.py:TestNonzero` so it's with the others

File: numpy/core/tests/test_numeric.py [hunk 6]
```diff
@@ -1175,6 +1175,35 @@ def __nonzero__(self):
...
+        assert_raises(RuntimeError, np.nonzero, a)
+
+        a = np.array([[True], [FalseThenTrue()]])
+        assert_raises(RuntimeError, np.nonzero, a)
+
+        a = np.array([False, TrueThenFalse()])
+        assert_raises(RuntimeError, np.nonzero, a)
+
+        a = np.array([[False], [TrueThenFalse()]])
+        assert_raises(RuntimeError, np.nonzero, a)
```
Comment (eric-wieser) [inline numpy/core/tests/test_numeric.py]: nit: PEP8 wants an extra blank line here

File: numpy/core/tests/test_numeric.py [hunk 7]
```diff
@@ -1175,6 +1175,35 @@ def __nonzero__(self):
...
+
+        class TrueThenFalse:
+            _val = True
+            def __bool__(self):
+                try:
+                    return self._val
+                finally:
+                    self._val = False
+
+        a = np.array([True, FalseThenTrue()])
```
Comment (eric-wieser) [inline numpy/core/tests/test_numeric.py]: nit: could do with a comment saying "grows on the second iteration"

File: numpy/core/tests/test_numeric.py [hunk 8]
```diff
@@ -1175,6 +1175,35 @@ def __nonzero__(self):
...
+        assert_raises(RuntimeError, np.nonzero, a)
+
+        a = np.array([[True], [FalseThenTrue()]])
+        assert_raises(RuntimeError, np.nonzero, a)
+
+        a = np.array([False, TrueThenFalse()])
```
Comment (eric-wieser) [inline numpy/core/tests/test_numeric.py]: nit: could do with a comment saying "shrinks on the second iteration"

**Review by eric-wieser — APPROVED** [review]
Just nits left I'd be happy to address myself via the github UI, otherwise looks good. Thanks for fixing this!

File: numpy/core/src/multiarray/item_selection.c (hunk 1 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/item_selection.c]: > I think we may as well error at too few items read.

I was going to suggest it, and was surprised to see it was already fixed, so thanks for noticing!

**Comment by garyg1** [comment]
Thanks @eric-wieser  and @seberg for the helpful feedback.

**Comment by garyg1** [comment]
@eric-wieser and sure, please do fix the nits if you want.

**Comment by seberg** [comment]
Thanks @garyg1 and @eric-wieser. Btw. I like the `gh-...` spelling, it comes from 8 years ago when we were moving, but then why not be specific that its github.

I suppose in principle this dance is not necessary for build-in dtypes which do not include references. But I doubt that it is noticable speed wise, so no point.

**Pull request merged**